### PR TITLE
feat: Overhaul the label to be set up according to the wanted implementations

### PR DESCRIPTION
**Pull request by EHandtke**
Overhauled the label to ensure all necessary methods are implemented. See below for demo code used to test the added label methods.

```java
package org.dwcj.component.label;

import org.dwcj.App;
import org.dwcj.component.TextAlignable;
import org.dwcj.component.TextAlignable.Alignment;
import org.dwcj.component.window.Frame;
import org.dwcj.component.window.Panel;
import org.dwcj.exceptions.DwcjException;

/**
 * Demos all the method of the label and consolelogs wether or not they complete correctly.
 */
public class LabelDemo extends App {
  @Override
  public void run() throws DwcjException {
    Frame page = new Frame();
    Panel div = new Panel();
    page.add(div);

    Label label = new Label("test");
    // div.add(label);

    div.setStyle("padding", "15px");
    div.setStyle("margin", "15px");
    div.setStyle("border", "solid black 1px");
    div.setStyle("width", "fit-content");
    div.setStyle("display", "flex");
    div.setStyle("flex-direction", "column");

    label.onMouseEnter(e -> App.consoleLog("label mouse entered"));
    label.onMouseExit(e -> App.consoleLog("label mouse exited"));
    label.onRightMouseDown(e -> App.consoleLog("label right mouse down"));

    App.consoleLog("Setting text for the label:");
    label.setText("newTest");
    App.consoleLog("Text after setting: " + label.getText());
    App.consoleLog("");

    App.consoleLog("Setting the label invisible:");
    label.setVisible(false);
    App.consoleLog("Label Visibility after setting: " + label.isVisible());
    App.consoleLog("");
    label.setVisible(true);

    App.consoleLog("Setting label disabled:");
    label.setEnabled(false);
    App.consoleLog("Enabled after setter: " + label.isEnabled());
    label.setEnabled(true);
    App.consoleLog("");

    App.consoleLog("Setting tooltiptext to Test:");
    label.setTooltipText("Test");
    App.consoleLog("Tooltiptext after setter: " + label.getTooltipText());
    App.consoleLog("");

    App.consoleLog("Setting the attribute test to testValue");
    label.setAttribute("test", "testValue");
    App.consoleLog("Attribute test after setter: " + label.getAttribute("test"));
    App.consoleLog("");

    App.consoleLog("Setting the style color to grey");
    label.setStyle("color", "grey");
    App.consoleLog("Style color after setter: " + label.getComputedStyle("color"));
    App.consoleLog("");

    App.consoleLog("Setting linewrap to false");
    label.setWrap(false);
    App.consoleLog("LineWrap after setter: " + label.isWrap());
    App.consoleLog("");

    App.consoleLog("Setting textAlignment to right");
    label.setTextAlignment(Alignment.RIGHT);
    App.consoleLog("TextAlignment after setter: " + label.getTextAlignment());
    App.consoleLog("");
    /*
     * Comment in for testing the catchup.
     */
    div.add(label);

  }
}
```

### CODE REVIEW

**Comment by mhawkinsbasis** [comment]
Two changes to be made:

1) Change the constructor to use the simpler signature of the underlying BBj control : addStaticText(string title, string flags)

2) **ONCE FOCUSABLE HAS BEEN IMPLEMENTED** - override the setFocusable() method for chaining.

**Comment by mhawkinsbasis** [comment]
Javadoc comments look good!

File: dwcj-engine/src/main/java/org/dwcj/component/label/Label.java [hunk 1]
```diff
@@ -3,59 +3,218 @@
...
+    return this;
   }
 
+  /**
+   * Sets whether the lines will be wrapped in the label.
+   *
+   * @param wrap - Specifies whether the lines will be wrapped (false = Not Wrapped, true = Wrapped)
+   * @return Returns this
+   */
+  public Label setLineWrap(Boolean wrap) {
```
Comment (hyyan) [inline dwcj-engine/src/main/java/org/dwcj/component/label/Label.java:188]: rename to setLineWrap

File: dwcj-engine/src/main/java/org/dwcj/component/label/Label.java [hunk 2]
```diff
@@ -3,59 +3,218 @@
...
+    return this;
+  }
+
+  /**
+   * Returns whether lines are wrapped in the label.
+   *
+   * @return Returns whether the lines are wrapped in the component (false = Not Wrapped, true =
+   *         Wrapped).
+   */
+  public Boolean isLineWrap() {
```
Comment (hyyan) [inline dwcj-engine/src/main/java/org/dwcj/component/label/Label.java:206]: rename to isWrap

File: dwcj-engine/src/main/java/org/dwcj/component/label/Label.java [hunk 3]
```diff
@@ -3,59 +3,218 @@
...
     this("");
   }
 
   /**
-   * Constructor used to give the label initial text
+   * Constructor used to give the label initial text.
    *
    * @param text String value for initial display text
    */
   public Label(String text) {
```
Comment (hyyan) [inline dwcj-engine/src/main/java/org/dwcj/component/label/Label.java:40]: Add a construct which accepts wrap as param

**Review by hyyan — CHANGES_REQUESTED** [review]

File: dwcj-engine/src/main/java/org/dwcj/component/label/Label.java [hunk 4]
```diff
@@ -3,59 +3,218 @@
...
     setText(text);
-    this.textAlignment = Alignment.LEFT;
+
```
Comment (hyyan) [inline dwcj-engine/src/main/java/org/dwcj/component/label/Label.java:42]: Remove empty line

**Review by hyyan — CHANGES_REQUESTED** [review]

File: dwcj-engine/src/main/java/org/dwcj/component/label/Label.java [hunk 5]
```diff
@@ -64,67 +223,113 @@ public Label setTextAlignment(Alignment alignment) {
...
   public Label setAttribute(String attribute, String value) {
     super.setAttribute(attribute, value);
     return this;
   }
 
+  /**
+   * {@inheritDoc}
+   */
   @Override
   public Label setId(String elementId) {
```
Comment (hyyan) [inline dwcj-engine/src/main/java/org/dwcj/component/label/Label.java:283]: Remove the method

**Review by hyyan — CHANGES_REQUESTED** [review]

File: dwcj-engine/src/main/java/org/dwcj/component/label/Label.java [hunk 6]
```diff
@@ -3,59 +3,237 @@
...
     setText(text);
-    this.textAlignment = Alignment.LEFT;
   }
 
+  /**
+   * Constructor used to give the label wether it is linewrapped or not.
+   *
+   * @param wrap Boolean value for linewrapping.
+   */
+  public Label(boolean wrap) {
```
Comment (hyyan) [inline dwcj-engine/src/main/java/org/dwcj/component/label/Label.java:49]: The constructor should accept both text and wrap

**Review by hyyan — CHANGES_REQUESTED** [review]

File: dwcj-engine/src/main/java/org/dwcj/component/label/Label.java [hunk 7]
```diff
@@ -3,59 +3,237 @@
...
+   * {@inheritDoc}
+   */
   @Override
   public Label setTextAlignment(Alignment alignment) {
     if (this.ctrl != null) {
       try {
-        ((BBjStaticText) ctrl).setAlignment(alignment.textPosition);
+        component.setAlignment(alignment.textPosition);
       } catch (BBjException e) {
         Environment.logError(e);
```
Comment (hyyan) [inline dwcj-engine/src/main/java/org/dwcj/component/label/Label.java:238]: Don't suppress errors. Throw a DWC run time exception

**Review by hyyan — CHANGES_REQUESTED** [review]

File: dwcj-engine/src/main/java/org/dwcj/component/label/Label.java (hunk 6 above)
Comment (EHandtke) [reply dwcj-engine/src/main/java/org/dwcj/component/label/Label.java:49]: I added two constructors, one for just wrap which will use default empty string for text and one for both.

**Comment by hyyan** [comment]
@EHandtke improve the posted test code to include wrap and text alignment.

File: dwcj-engine/src/main/java/org/dwcj/component/label/Label.java [hunk 8]
```diff
@@ -3,122 +3,332 @@
...
-                                  // of checks
   protected void catchUp() throws IllegalAccessException {
-    if (Boolean.TRUE.equals(this.getCaughtUp()))
+    if (Boolean.TRUE.equals(this.getCaughtUp())) {
       throw new IllegalAccessException("catchUp cannot be called twice");
+    }
     super.catchUp();
-  }
 
+    if (!this.lineWrap) {
```
Comment (hyyan) [inline dwcj-engine/src/main/java/org/dwcj/component/label/Label.java:326]: This will fail if lineWrap is false

**Review by hyyan — CHANGES_REQUESTED** [review]

File: dwcj-engine/src/main/java/org/dwcj/component/label/Label.java [hunk 9]
```diff
@@ -3,116 +3,325 @@
 import com.basis.bbj.proxies.sysgui.BBjStaticText;
 import com.basis.bbj.proxies.sysgui.BBjWindow;
 import com.basis.startup.type.BBjException;
-
-import org.dwcj.Environment;
 import org.dwcj.bridge.WindowAccessor;
 import org.dwcj.component.AbstractDwcComponent;
 import org.dwcj.component.HasEnable;
```
Comment (hyyan) [inline dwcj-engine/src/main/java/org/dwcj/component/label/Label.java:8]: remove unrequired import

File: dwcj-engine/src/main/java/org/dwcj/component/label/Label.java [hunk 10]
```diff
@@ -3,116 +3,325 @@
...
-   * with this
-   */
+  private EventDispatcher dispatcher = new EventDispatcher();
+  private MouseEnterEventSink mouseEnterEventSink;
+  private MouseExitEventSink mouseExitEventSink;
+  private RightMouseDownEventSink rightMouseDownEventSink;
+  private boolean lineWrap = true;
+  private BBjStaticText component;
 
   public Label() {
```
Comment (hyyan) [inline dwcj-engine/src/main/java/org/dwcj/component/label/Label.java:33]: Add javadocs

File: dwcj-engine/src/main/java/org/dwcj/component/label/Label.java [hunk 11]
```diff
@@ -3,116 +3,325 @@
...
-  /*
-   * top, right, bottom, left margins, used simple array as user will not have to interact directly
-   * with this
-   */
+  private EventDispatcher dispatcher = new EventDispatcher();
+  private MouseEnterEventSink mouseEnterEventSink;
+  private MouseExitEventSink mouseExitEventSink;
+  private RightMouseDownEventSink rightMouseDownEventSink;
+  private boolean lineWrap = true;
+  private BBjStaticText component;
```
Comment (hyyan) [inline dwcj-engine/src/main/java/org/dwcj/component/label/Label.java:31]: remove and use `this.ctrl`

File: dwcj-engine/src/main/java/org/dwcj/component/label/Label.java [hunk 12]
```diff
@@ -3,116 +3,338 @@
...
+   */
   @Override
-  protected void create(AbstractWindow p) {
+  protected void create(AbstractWindow p) throws DwcjRuntimeException {
     try {
       BBjWindow w = WindowAccessor.getDefault().getBBjWindow(p);
       byte[] flags = BBjFunctionalityHelper.buildStandardCreationFlags(this.isVisible(), true);
-      ctrl = w.addStaticText(w.getAvailableControlID(), BASISNUMBER_1, BASISNUMBER_1, BASISNUMBER_1,
-          BASISNUMBER_1, getText(), flags);
+      ctrl = w.addStaticText(getText(), flags);
```
Comment (hyyan) [inline dwcj-engine/src/main/java/org/dwcj/component/label/Label.java:65]: Format the document

**Review by hyyan — APPROVED** [review]

**Pull request merged**